### ISSUE TITLE
Cache national cost surface data to speed up usage

### ISSUE THREAD

**Issue by machristie** [SciGaP/simccs-maptool#69]
## TODO

- [x] don't load the national cost surface data on start up, load on first use?
- [x] submit PR to SimCCS for new getters

**Comment by machristie** [comment]
Created a CostSurfaceData class in the SimCCS to hold the three arrays, constructionCosts, rightOfWayCosts, and routingCosts. I can get this out of a DataStorer instance that has called loadNetworkCosts(), then set this on a new DataStorer instance so it doesn't need to load it from files.

What I'm finding is that it takes about 10-20 seconds to set CostSurfaceData on a new DataStorer instance. That's better than the ~2 minutes it takes to load cost surface data from disk.

However, generating a candidate graph is still taking about 4 minutes, at least on my laptop. So maybe loading the cost surface data from disk isn't the main component of the cost of generating a candidate graph or mps file.

I want to test this on the server to see if I see the same timings there.

**Comment by machristie** [comment]
Here's the python script I'm using to test:

```python
# coding: utf-8
from datetime import datetime
import jnius_config
import logging
import os
import sys


logging.basicConfig(level=logging.DEBUG, format="%(asctime)s %(levelname)s %(name)s: %(message)s")
logger = logging.getLogger(__name__)

os.environ['JAVA_HOME'] = '/Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home'
# jnius_config.set_classpath(os.path.join(os.getcwd(), 'store', 'SimCCS.jar'))
jnius_config.set_classpath(
    os.path.join(os.getcwd(), "store", "SimCCS.jar"),
)
jnius_config.add_options('-Xmx12G')
from jnius import autoclass

dataset, scenario = "Lower48US", "scenario1"
basepath = os.path.join(os.getcwd(), "Datasets")

DataStorer = autoclass("dataStore.DataStorer")
Solver = autoclass("solver.Solver")

def load_cost_surface_data():

    data = DataStorer(basepath, dataset, scenario)
    solver = Solver(data)
    data.setSolver(solver)
    logger.debug("loading network costs")
    data.loadNetworkCosts()
    logger.debug("finished loading network costs")
    return data.getCostSurfaceData()

cost_surface_data = load_cost_surface_data()
logger.debug("finished loading cost_surface_data")

try:
    data = DataStorer(basepath, dataset, scenario)
    solver = Solver(data)
    data.setSolver(solver)
    logger.debug("setting cost_surface_data")
    data.setCostSurfaceData(cost_surface_data)
    logger.debug("generating candidate graph")
    data.generateCandidateGraph()
except Exception as e:
    logger.exception(
        "Error: " + str(e.stacktrace)
        if hasattr(e, "stacktrace")
        else str(e)
    )
    raise

logger.info("finished")
```

**Comment by machristie** [comment]
Maybe use https://pyjnius.readthedocs.io/en/stable/api.html#java-class-implementation-in-python to implement the CostSurfaceData class in Python? SimCCS codebase doesn't really have much use for it.

**Comment by machristie** [comment]
Here are the timings on the beta server:
```
$ python cost_surface_data.py 
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
2020-04-21 13:47:49,903 DEBUG __main__: loading network costs
2020-04-21 13:50:11,089 DEBUG __main__: finished loading network costs
2020-04-21 13:50:11,091 DEBUG __main__: finished loading cost_surface_data
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
2020-04-21 13:50:11,097 DEBUG __main__: setting cost_surface_data
2020-04-21 13:50:16,470 DEBUG __main__: generating candidate graph
2020-04-21 13:51:21,078 INFO __main__: finished
```

So it takes 2:22 minutes to load the cost surface data, then 5 seconds to set on a second scenario, then 1:05 minutes to generate the candidate graph.

On a second run, these times are 2:56, 0:01 and 1:21 respectively.

On a third run, these times are 1:52, 0:09 and 1:02 respectively.

---

Now I'll compare against the standard approach, that is, **without the cost surface data cache**.

```
$ python cost_surface_data.py 
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
2020-04-21 14:14:25,866 DEBUG __main__: generating candidate graph
2020-04-21 14:17:56,006 INFO __main__: finished
```

This takes 3:31 to load cost surface + generate candidate graph.

On a second run it takes 2:54.

On a third run it takes 2:51 minutes.

---

So in summary:
- without cached cost surface, generating a candidate network takes on average 3:05 minutes
- with cached cost surface, generating a candidate network takes on average 1:09 minutes
- that's about a 63% reduction in time to generate a candidate network

**Comment by machristie** [comment]
> Maybe use https://pyjnius.readthedocs.io/en/stable/api.html#java-class-implementation-in-python to implement the CostSurfaceData class in Python? SimCCS codebase doesn't really have much use for it.

Turns out this won't really work. For one, it needs an interface to implement. And secondly, this just creates a Java proxy to the Python implementation, meaning that I would likely have the same problem when cost surface arrays are converted back and forth between Python and Java runtimes.

**Comment by machristie** [comment]
Profiling the loading of cost surface data. I created a [Main.java](https://github.com/machristie/SimCCS/blob/cost-loading-timings/src/Main.java) to just load network costs. Also on this branch I added some timing and print statements to break down the cost of various operations: https://github.com/machristie/SimCCS/tree/cost-loading-timings

In consultation with Sean I found out that the `RightOfWay Costs.txt` file is no longer used, so I removed that from the Lower48US BaseData for these timings.

Here's how to build the code:

```
javac -cp store/SimCCS.jar -d build src/Main.java src/dataStorer/DataInOut.java
```

Here are the timings I get:

### 12GB max heap
```
java -Xmx12G -cp build:store/SimCCS.jar Main
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
Loading ./Datasets/Lower48US/BaseData/CostNetwork/Construction Costs.csv
Load construction csv file: 29.616740639 s
Setup routing costs array: 10.324265017 s
Populate routing costs array: 0.480894999 s
data.setX() method calls: 17.767279824 s
```

Here are other timings that I tried with different heap sizes. These all ran on my laptop (MacBook Pro (15-inch, 2016), 2.6 GHz Intel Core i7, 16 GB 2133 MHz LPDDR3).

### 8GB max heap
```
java -Xmx8G -cp build:store/SimCCS.jar Main
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
Loading ./Datasets/Lower48US/BaseData/CostNetwork/Construction Costs.csv
Load construction csv file: 29.191400372 s
Setup routing costs array: 7.411701654 s
Populate routing costs array: 0.387350115 s
data.setX() method calls: 10.559467387 s
```

### 6GB max heap
Note: didn't complete, I eventually killed it. This probably isn't enough heap space.

```
java -Xmx6G -cp build:store/SimCCS.jar Main
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
Loading ./Datasets/Lower48US/BaseData/CostNetwork/Construction Costs.csv
Load construction csv file: 27.511238664 s
Setup routing costs array: 7.446713967 s
Populate routing costs array: 0.395894622 s
^CKilled: 9
```

### 10 GB max heap
```
java -Xmx10G -cp build:store/SimCCS.jar Main
Loading Geography...
Loading Source Data...
Loading Sink Data...
Loading Transport Data...
Loading Delaunay Pairs...
Loading Candidate Graph...
Data Loaded.
Loading ./Datasets/Lower48US/BaseData/CostNetwork/Construction Costs.csv
Load construction csv file: 27.938466504 s
Setup routing costs array: 10.161046174 s
Populate routing costs array: 0.399644007 s
data.setX() method calls: 13.444494788 s
```
